# WARNING - Generated by {fusen} from dev/flat_first.Rmd: do not edit by hand

#' origene_primer_search
#' 
#' origene_primer_search for primer design 
#' 
#' @param gene str gene smbol
#' @param speice str default as Human
#' @param ncbi_id the ncbi id of gene,default as NULL
#' @param use_ncbi use ncbi database when True,default as False
#' @return vector
#' 
#' @export
#' @examples
#' origene_primer_search(gene = "itga9",
#'                                       speice = "Human",
#'                                       use_ncbi = F)
origene_primer_search <- function(gene = "ITGA9",
                                  speice = "Human",
                                  ncbi_id = NULL,
                                  use_ncbi = F){
  message("loading packages")
  #library(dplyr)
  #library(rvest)
  #library(stringr)
  #library(glue)
  message("searching gene in origene webpage ...")
  origene_search <- glue::glue("https://www.origene.com.cn/search?q={gene}&species_filter={speice}&model=&sub_category=qPCR+Primer+Pairs&fromPage=ivd")
  origene_html <- origene_search %>% 
    rvest::read_html()
  dest_link <- origene_html %>% 
    rvest::html_elements("a.name") %>% 
    rvest::html_attr("href")
  if (length(dest_link) == 0 | use_ncbi ) {
    message("Gene not found in origene, try search in NCBI")
    if (is.null(ncbi_id)){
      ncbi_html <- glue::glue("https://www.ncbi.nlm.nih.gov/gene/?term={gene}") %>% 
        rvest::read_html() 
      ncbi_id <- ncbi_html %>% 
        rvest::html_elements("span.gene-id") %>% 
        rvest::html_text() %>% 
        stringr::str_remove(.,"ID:") %>% 
        stringr::str_trim(.)
      ncbi_speice <- ncbi_html %>% 
        rvest::html_elements("em") %>% 
        rvest::html_text() 
      ncbi_speice <- ncbi_speice[1:length(ncbi_id)]
      if (speice == "Human") {
        speice = "Homo sapiens"
      }else if (speice == "Mouse") {
        speice = "Mus musculus"
      } else {
        stop(glue::glue("input a speice from the following list: {paste(ncbi_speice,sep = ',')} "))
      }
      ncbi_id <- ncbi_id[grep(speice,ncbi_speice)]
      message(glue::glue("processing Gene {gene} with NCBI ID:{ncbi_id}"))
    }
    message("loading package chromote, please make sure that you had Chrome browser installed ")
    library(chromote)
    ncbi_transcript <- glue::glue("https://www.ncbi.nlm.nih.gov/datasets/gene/id/{ncbi_id}/products/") %>% 
      rvest::read_html_live() %>% 
      rvest::html_elements("td") %>% 
      rvest::html_text()
    ncbi_transcript_NM <- ncbi_transcript[grepl("NM_",ncbi_transcript)]
    message(glue::glue("Detect transcript {paste(ncbi_transcript_NM,sep = ',')}"))
    message("Further Primer design at https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi?LINK_LOC=BlastHome")
    return(NULL)
  }else {
    message("fetching gene information ...")
    dest_info <-  paste0("https://www.origene.com.cn",dest_link[1])%>% 
      rvest::read_html() %>% 
      rvest::html_elements("tr.attribute") %>% 
      rvest::html_text()
    forward_sequence = dest_info[grepl("Forward Sequence",dest_info)] %>% 
      stringr::str_remove(.,"Forward Sequence\n") %>% 
      stringr::str_remove_all(.,"\\s+") %>% 
      stringr::str_remove(.,"\n")
    message(glue::glue("forward_sequence is {forward_sequence}"))
    reverse_sequene = dest_info[grepl("Reverse Sequence",dest_info)] %>% 
      stringr::str_remove(.,"Reverse Sequence\n") %>% 
      stringr::str_remove_all(.,"\\s+") %>% 
      stringr::str_remove(.,"\n")
    message(glue::glue("reverse_sequene is {reverse_sequene}"))
    message("now validate the above sequences at https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi?LINK_LOC=BlastHome")
    primer <- c(forward_sequence,reverse_sequene)
    names(primer) <- c("forward_sequence","reverse_sequene")
    return(primer)
  }
}

